File: examples/documents/upsert-by-unique.ts
import { Category } from "../_harness/generated/ts/Category.generated";

// Upsert by a NAMED unique constraint (single- or multi-field). Finds an
// existing record by the constraint and merges the data into it, or creates a
// new record if none matches. The lookup values are passed in field order, and
// the data object MUST carry the same constraint-field values (a mismatch
// throws). `targetDocument` is required whenever a new record may be created.
export async function upsertCategory(documentId: string) {
  // #region example
  // "name_parentId" is the named constraint declared in the schema
  // ([[models.categories.unique_constraints]] name = "name_parentId",
  //  fields = ["name", "parentId"]).
  const category = await Category.upsertByUnique(
    "name_parentId", // the constraint NAME — not the field list
    ["Work", "root"], // lookup values, in the constraint's field order
    { name: "Work", parentId: "root", color: "blue" },
    { targetDocument: documentId }, // required if a new record is created
  );
  // #endregion example
  return category;
}

File: examples/documents/upsert-by-unique.swift
import Foundation
import JsBaoClient

// Upsert by a NAMED unique constraint (single- or multi-field). The lookup
// values come straight from the record's own fields — every constraint field
// must be set. `upsertByUnique` merges into the existing record matched by the
// constraint, or creates a new one if none matches, returning the resolved
// record (on the merge path its `id` is the existing record's id).
func upsertCategory(documentId: String) throws {
  // #region example
  let category = Category(
    id: UUID().uuidString,
    name: "Work",
    parentId: "root",
    color: "blue"
  )
  // "name_parentId" is the named constraint declared in models.toml. `mode`
  // defaults to .either (create-or-update); pass .mustExist / .mustNotExist to
  // require a single path.
  let resolved = try category.upsertByUnique("name_parentId", in: documentId)
  // #endregion example
  _ = resolved
}
